File: legacy/src/game/pinball-knight/sim/paused.ts
/**
 * THE PAUSE CONTRACT.
 *
 * True while ANY modal surface owns the screen. `simulate` early-returns on it,
 * and the frame loop books the elapsed wall-clock into `state.pausedRunS` so the
 * leaderboard's run duration does not count time spent reading.
 *
 * It used to ask whether four particular DOM nodes existed — so the most
 * consequential boolean in the game lived in the document tree. Screens are
 * values on a stack now, and `gui/stack.ts` maintains `state.uiPauses` from the
 * screens that declare they pause. The walkable tavern is still its own scene
 * and answers separately.
 *
 * Its own module because both halves of the frame — the simulation and the loop
 * — ask it, and it must not drag either of them back into core.
 */
import { state } from "../state";
import { isTavernSceneOpen } from "../../../scenes/tavern";

export function isSimPaused(): boolean {
  return state.uiPauses || isTavernSceneOpen();
}

File: Rust/crates/pk-core/src/sim/paused.rs
//! The Simulation Pause Contract — Universal modal pause check for the physics tick and frame loop.
//!
//! PORTS: `sim/paused.ts`

#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub struct PauseState {
    pub ui_pauses: bool,
    pub tavern_open: bool,
}

impl PauseState {
    pub fn new(ui_pauses: bool, tavern_open: bool) -> Self {
        Self {
            ui_pauses,
            tavern_open,
        }
    }

    /// True while ANY modal surface or tavern scene owns the screen.
    pub fn is_sim_paused(&self) -> bool {
        self.ui_pauses || self.tavern_open
    }
}

/// Evaluates whether the simulation should pause based on active modal UI or tavern state.
pub fn is_sim_paused(ui_pauses: bool, tavern_open: bool) -> bool {
    ui_pauses || tavern_open
}
